### PR TITLE
[sharing] fix addToGroup hook

### PR DESCRIPTION
**Pull request by schiessle**
Fix the add to group hook in sharing. This was discovered while working on https://github.com/owncloud/core/pull/17327

We skip this step completely for file shares because for file shares we generate a unique target during the mount operation if needed. Further we make sure that the code gets executed at all. In the old version "$sourcrExists" was always true because we checked it in a post hook. So we always found a share for the user after he was added to the group.

cc @rullzer maybe you have some time to review it? Thanks!

### CODE REVIEW

**Comment by schiesbn** [comment]
cc @blizzz maybe you can also have a look as we discussed the issue earlier today.

File: lib/private/share/hooks.php [hunk 1]
```diff
@@ -47,39 +47,59 @@ public static function post_deleteUser($arguments) {
...
+						'`parent`' => $insert->expr()->literal($item['id']),
+						'`share_type`' => $insert->expr()->literal(self::$shareTypeGroupUserUnique),
+						'`share_with`' => $insert->expr()->literal($arguments['uid']),
+						'`uid_owner`' => $insert->expr()->literal($item['uid_owner']),
+						'`permissions`' => $insert->expr()->literal($item['permissions']),
+						'`stime`' => $insert->expr()->literal($item['stime']),
+						'`file_source`' => $insert->expr()->literal($item['file_source']),
+					]
+				);
+				$insert->execute();
```
Comment (rullzer) [inline lib/private/share/hooks.php]: Does this work correctly when you run execute multiple times? Since you keep the original $insert object.

**Comment by rullzer** [comment]
Looks good. But I would really like unit tests here.
It is a bit unintuitive that we do this stuff somewhere else for files...

**Comment by schiesbn** [comment]
> It is a bit unintuitive that we do this stuff somewhere else for files...

For files we need to check anyway on every mount for potential name collisions and resolve it if it happens. Because on a external storage someone could create a file with the same name, the next time you login to you ownCloud the share needs to be moved to a unique name. 

Also I think it is better for performance reasons on a per user basis if needed instead of checking it for all users during the share operation. Most likely I would remove this logic completely from a share API 2.0 and let the back-end handle it, like it is already done by sharing.

File: lib/private/share/hooks.php (hunk 1 above)
Comment (schiesbn) [inline lib/private/share/hooks.php]: should work because we replace the values on the object for each execute()... but I will double-check.

**Comment by schiesbn** [comment]
Hm, tests fail for the same reason it didn't worked for files. We are in a post-hook. This means that the user is already a member of the group. If we now ask for all shares we will already get the new group share, detect a collision and generate a new target "test1.txt". I don't see a way to fix this for the tests. Don't know why this works for address books... will look deeper into it.

**Comment by schiesbn** [comment]
@rullzer please have another look, I changed the approach based on https://github.com/owncloud/core/pull/17381#issuecomment-118784477

We now have a pre-hook to calculate the unique target name and a post-hook to add it to the table

**Comment by rullzer** [comment]
@schiesbn I don't have a 7.0 install... Code looks good to me. But I can't verify it :P

**Comment by rullzer** [comment]
Ignore that last one... I'll have another look :)

File: lib/private/share/hooks.php [hunk 2]
```diff
@@ -41,46 +48,96 @@ public static function post_deleteUser($arguments) {
...
 
+
 			// Insert an extra row for the group share if the item or file target is unique for this user
-			if ($itemTarget != $item['item_target'] || $fileTarget != $item['file_target']) {
-				$query->execute(array($item['item_type'], $item['item_source'], $itemTarget, $item['id'],
-					self::$shareTypeGroupUserUnique, $arguments['uid'], $item['uid_owner'], $item['permissions'],
-					$item['stime'], $item['file_source'], $fileTarget));
-				\OC_DB::insertid('*PREFIX*share');
+			if (($fileTarget === null && $itemTarget != $item['item_target']) ||
+				($fileTarget !== null && $fileTarget !== $item['file_target'])) {
```
Comment (blizzz) [inline lib/private/share/hooks.php]: coding style: please put || at the beginning of the new line

**Comment by blizzz** [comment]
The code itself looks good, and also I really appreciate the tests. Is there anything specific to test manually (steps)?

**Comment by schiesbn** [comment]
Following steps can be used to test the correct behavior:

- create admin, user2, user3 and user4
- add user2 to group "users"
- admin shares a file test.txt with the group "users"
- login as user3
- user3 also creates a test.txt
- admin adds user3 to the group "users"

Now check the oc_share table in the database. With this patch you should see a extra row for the group share for user3 with a target "test.txt (2)" or similar because we created a unique name for him.

- admin adds user4 to the group "users"

Now check the oc_share table in the database. No additional row for the group share for user4 should be created.

Note: It is important that you check the db directly after adding user3 (user4) to the group. If user3 log-in we will detect the name collision and create a unique target for him. So after login everything will look the same with or without the patch.

File: lib/private/share/hooks.php (hunk 2 above)
Comment (schiesbn) [inline lib/private/share/hooks.php]: done

**Comment by rullzer** [comment]
I assume this has to go into 8.2?

**Comment by rullzer** [comment]
Code looks good
I also checked your example case and that works as expected!

Probably hard to add that tests to the testsuite?

Otherwise :+1:

**Comment by schiesbn** [comment]
@MorrisJobke not sure if it is worth a backport. Because it affects mainly calendar and contact. For files we create the unique target on mount if needed.

**Comment by schiesbn** [comment]
needs a second reviewer... @nickvergessen ? Thanks!

**Comment by MorrisJobke** [comment]
:+1:

**Comment by MorrisJobke** [comment]
@schiesbn What about the backport to stable8.1? This was only merged against master.

**Comment by MorrisJobke** [comment]
@schiesbn Ping

**Comment by MorrisJobke** [comment]
@schiesbn You filed this ticket for 8.1.1 and this is against master, so you need to create a backport.

**Comment by MorrisJobke** [comment]
* [ ] if this is backported then also backport #18876

**Pull request merged**